sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateJoinByCombineAggregate.scala

```diff
@@ -0,0 +1,196 @@
+                  ae.copy(filter = Some(filters.last))
+              }.asInstanceOf[NamedExpression]
+            }
+          } else {
+            la.aggregateExpressions ++ rightAggregateExprs
+          }
+
+          (Aggregate(Seq.empty, mergedAggregateExprs, newChild), Map.empty, Seq.empty)
+        }
+      case (lp: Project, rp: Project) =>
+        val mergedInfo = mergePlan(lp.child, rp.child)
+        val mergedProjectList = ArrayBuffer[NamedExpression](lp.projectList: _*)
+
+        mergedInfo.map { case (newChild, outputMap, filters) =>
+          val allFilterReferences = filters.flatMap(_.references)
+          val newOutputMap = (rp.projectList ++ allFilterReferences).map { ne =>
+            val mapped = ne.transform {
+              case attr: Attribute =>
+                outputMap.getOrElse(attr.canonicalized, attr)
+            }.asInstanceOf[NamedExpression]
+
+            val withoutAlias = mapped match {
+              case Alias(child, _) => child
+              case e => e
+            }
+
+            val outputAttr = mergedProjectList.find {
+              case Alias(child, _) => child semanticEquals withoutAlias
+              case e => e semanticEquals withoutAlias
+            }.getOrElse {
+              mergedProjectList += mapped
+              mapped
+            }.toAttribute
+            ne.toAttribute.canonicalized -> outputAttr
+          }.toMap
+
+          (Project(mergedProjectList.toSeq, newChild), newOutputMap, filters)
+        }
+      // TODO support only one side contains Filter
+      case (lf: Filter, rf: Filter) =>
+        val mergedInfo = mergePlan(lf.child, rf.child)
+        mergedInfo.map { case (newChild, outputMap, _) =>
+          val rightCondition = rf.condition transform {
+            case attr: Attribute =>
+              outputMap.getOrElse(attr.canonicalized, attr)
+          }
+          val newCondition = Or(lf.condition, rightCondition)
+
+          (Filter(newCondition, newChild), outputMap, Seq(lf.condition, rightCondition))
+        }
+      case (ll: LeafNode, rl: LeafNode) =>
+        if (ll.canonicalized == rl.canonicalized) {
+          val outputMap = rl.output.zip(ll.output).map { case (ra, la) =>
+            ra.canonicalized -> la
+          }.toMap
+
+          Some((ll, outputMap, Seq.empty))
+        } else {
+          None
+        }
+      case (ls: SerializeFromObject, rs: SerializeFromObject) =>
+        if (ls.canonicalized == rs.canonicalized) {
+          val outputMap = rs.output.zip(ls.output).map { case (ra, la) =>
+            ra.canonicalized -> la
+          }.toMap
+
+          Some((ls, outputMap, Seq.empty))
+        } else {
+          None
+        }
+      case _ => None
+    }
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!conf.eliminateJoinByCombineAggregateEnabled) return plan
+
+    plan.transformDownWithPruning(_.containsAnyPattern(JOIN, AGGREGATE), ruleId) {
```

Comment:
I think transform up is better. For a multi-join query, we can eliminate the leaf joins first, and transform up to eliminate all joins if possible.